integration/middleware-test.ts

```diff
@@ -0,0 +1,396 @@
+
+            let res = await context.next();
+
+            count = context.get(countContext);
+            context.set(countContext, ++count)
+            res.headers.set('x-count', count);
+            return res;
+          }
+
+          export function action({ context }) {
+            let count = context.get(countContext);
+            return json({ count: count + 1 });
+          }
+
+          export function loader({ context }) {
+            return json({ count: context.get(countContext) + 1 });
+          }
+
+          export default function Component() {
+            let data = useLoaderData();
+            let actionData = useActionData();
+            return (
+              <div>
+                <p id="grandchild-data">{JSON.stringify(data)}</p>
+                <p id="grandchild-action-data">{actionData ? JSON.stringify(actionData) : 'empty'}</p>
+                <Form action={"?count=" + data.count} method="post">
+                  <button type="submit" name="count" value="100">Submit</button>
+                </Form>
+              </div>
+            )
+          }
+        `,
+      },
+    });
+
+    appFixture = await createAppFixture(fixture, ServerMode.Development);
+  });
+
+  test.afterAll(() => {
+    appFixture.close();
+  });
+
+  test.describe("without JavaScript", () => {
+    test.use({ javaScriptEnabled: false });
+    runMiddlewareTests();
+  });
+
+  test.describe("with JavaScript", () => {
+    test.use({ javaScriptEnabled: true });
+    runMiddlewareTests();
+  });
+
+  test.describe("Response Headers", () => {
+    // These middlewares increment on the way down, in actions/loaders, and on
+    // the way back up and set count in a response header
+
+    test("Allows middleware to alter loader responses (document requests)", async () => {
+      let response;
+
+      response = await fixture.requestDocument("/parent-response");
+      expect(response.headers.get("x-count")).toBe("2");
+
+      response = await fixture.requestDocument("/parent-response/child");
+      expect(response.headers.get("x-count")).toBe("4");
+
+      response = await fixture.requestDocument(
+        "/parent-response/child/grandchild"
+      );
+      expect(response.headers.get("x-count")).toBe("6");
+    });
+
+    test("Allows middleware to alter action responses (document requests)", async () => {
+      let response;
+
+      let formData = new FormData();
+      formData.set("count", "100");
+      response = await fixture.requestDocument(
+        "/parent-response/child/grandchild",
+        {
+          method: "post",
+          body: formData,
+        }
+      );
+      expect(response.headers.get("x-count")).toBe("106");
+    });
+
+    test("Allows middleware to alter loader responses (data requests)", async () => {
+      let response;
+
+      response = await fixture.requestData(
+        "/parent-response",
+        "routes/parent-response"
+      );
+      expect(response.headers.get("x-count")).toBe("2");
```

Comment:
These tests are OK, but I think you could be more explicit than a number. What about something like this?
```
expect(response.headers.get("x-processed-by")).toBe("parent")
…
expect(response.headers.get("x-processed-by")).toBe("parent,child")
…
expect(response.headers.get("x-processed-by")).toBe("parent,child,grandchild")
```

That makes it clearer which middleware are being invoked for each request.